### PR TITLE
SONARJAVA-2207 fix infinite recursion when resolving relations

### PR DESCRIPTION
**Pull request by saberduck**
- fixes SOE in elasticsearch analysis
 - contains also refactoring to reuse transitive relationships between resolution and copying of constraints

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationState.java [hunk 1]
```diff
@@ -39,16 +37,6 @@
    */
   UNDETERMINED;
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationState.java]: MAJOR Unknown constructor [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownConstructorCall)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [RelationState.java#L30](https://github.com/SonarSource/sonar-java/blob/f6c2282af387ce91d7cdbbd799c3e84a05f0645c/java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationState.java#L30): Unknown constructor [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownConstructorCall)
1. ![MAJOR][MAJOR] [RelationState.java#L34](https://github.com/SonarSource/sonar-java/blob/f6c2282af387ce91d7cdbbd799c3e84a05f0645c/java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationState.java#L34): Unknown constructor [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownConstructorCall)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java [hunk 2]
```diff
@@ -262,11 +244,28 @@ private RelationState implies(RelationalSymbolicValue relation) {
...
+    workList.add(this);
+    while (!workList.isEmpty()) {
+      if (newRelations.size() > MAX_DEDUCED_RELATIONS || iterations > MAX_ITERATIONS) {
+        // safety mechanism in case of an error in the algorithm
+        // should not happen under normal conditions
+        throw new RelationalSymbolicValue.TransitiveRelationExceededException("Used relations: " + newRelations.size() + ". Iterations " + iterations);
+      }
+      iterations++;
+      RelationalSymbolicValue relation = workList.pop();
+      for (RelationalSymbolicValue knownRelation : knownRelations) {
```
Comment (saberduck) [inline java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java]: I am hesitating to use Streams here, I tried and it didn't seem more legible to me. Also I find this easier to debug. wdyt?>

File: java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java [hunk 3]
```diff
@@ -101,16 +100,25 @@ public RelationalSymbolicValue(Kind kind) {
 
   @Override
   public List<ProgramState> setConstraint(ProgramState initialProgramState, BooleanConstraint booleanConstraint) {
-    if (!checkRelation(booleanConstraint, initialProgramState)) {
+    if (booleanConstraint == BooleanConstraint.FALSE) {
+      return inverse().setConstraint(initialProgramState, BooleanConstraint.TRUE);
+    }
+    Set<RelationalSymbolicValue> knownRelations = knownRelations(initialProgramState).collect(Collectors.toSet());
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java]: given that method knownRelations is only used here I don't see the point of making it returning a stream. Let it return a set. I am even wondering about making this method part of PS. But that's something else.

File: java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java [hunk 4]
```diff
@@ -197,41 +205,15 @@ RelationalSymbolicValue inverse() {
     return states;
   }
 
-  private boolean checkRelation(BooleanConstraint booleanConstraint, ProgramState programState) {
-    RelationState relationState = resolveState(programState);
-    return !relationState.rejects(booleanConstraint);
-  }
-
-  RelationState resolveState(ProgramState programState) {
+  RelationState resolveState(Set<RelationalSymbolicValue> knownRelations) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java]: lack VisibleForTesting annotation.
I know this vocabulary does not come from you but name of the method is confusing : resolveRelationState would be a bit more explicit (overall RelationState is a bad name IMO but let's not change it now)

File: java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java [hunk 5]
```diff
@@ -262,11 +244,28 @@ private RelationState implies(RelationalSymbolicValue relation) {
...
-      .collect(Collectors.toList());
+  private Set<RelationalSymbolicValue> transitiveRelations(Set<RelationalSymbolicValue> knownRelations) {
+    Set<RelationalSymbolicValue> newRelations = new HashSet<>();
+    Deque<RelationalSymbolicValue> workList = new ArrayDeque<>();
+    int iterations = 0;
+    workList.add(this);
+    while (!workList.isEmpty()) {
+      if (newRelations.size() > MAX_DEDUCED_RELATIONS || iterations > MAX_ITERATIONS) {
+        // safety mechanism in case of an error in the algorithm
+        // should not happen under normal conditions
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java]: this comments lack a bit of explanation about what "normal" is.

File: java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java [hunk 6]
```diff
@@ -262,11 +244,28 @@ private RelationState implies(RelationalSymbolicValue relation) {
...
+        throw new RelationalSymbolicValue.TransitiveRelationExceededException("Used relations: " + newRelations.size() + ". Iterations " + iterations);
+      }
+      iterations++;
+      RelationalSymbolicValue relation = workList.pop();
+      for (RelationalSymbolicValue knownRelation : knownRelations) {
+        RelationalSymbolicValue r = relation.deduceTransitiveOrSimplified(knownRelation);
+        if (r != null && !knownRelations.contains(r) && !newRelations.contains(r)) {
+          newRelations.add(r);
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java]: could be done slightly better with  : 
```
 if (r != null && !knownRelations.contains(r)) {
   if(newRelations.add(r)) {
      worklist.add(r);
   }
 }
```
and those two if can be in turn collapsed : Wonder about readability though...

**Comment by benzonico** [comment]
You only have indirect tests. So while I am pretty confident that you fixed the issue I struggle to see exactly what fixed what : you fixed transitive computation ? please add a dedicated test for it.

File: java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java (hunk 5 above)
Comment (saberduck) [inline java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java]: I removed the whole line, it's better without

File: java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java (hunk 6 above)
Comment (saberduck) [inline java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java]: I changed like suggested, though also not 100% convinced about having sideeffect in the if condition . Nested if raises warning about too many nested control flow statements.

**Pull request merged**